### PR TITLE
feat(generators): lower step resilience into the saga — retry/timeout/compensation/confidence + WithContext (#135)

### PR DESCRIPTION
**Pull request by rsalus**
## Summary

Fixes the core bug in **#135**: step-level resilience — `WithRetry` / `WithTimeout` / `Compensate<T>` / `RequireConfidence` + `OnLowConfidence` — was *declared, documented, and exported* but **never lowered into the emitted Wolverine + Marten saga** (a shipped, non-functional feature). It now lowers end-to-end, each capability **proven against a real Wolverine + Marten + Postgres host** (not shape-only). `WithContext` (the dead `ContextAssemblerEmitter`) is wired in too.

Design: [`docs/designs/2026-06-17-step-resilience-lowering.md`](docs/designs/2026-06-17-step-resilience-lowering.md) · Plan: [`docs/plans/2026-06-17-step-resilience-lowering.md`](docs/plans/2026-06-17-step-resilience-lowering.md)

## What now lowers (each with a real-container behavioral test)

| DR | Capability | Mechanism | Proof |
|----|-----------|-----------|-------|
| DR-2 | **Retry** | per-handler `Configure(HandlerChain)` → `OnAnyException().RetryTimes`/`RetryWithCooldown` | flaky step retried exactly 3× → completes |
| DR-3 | **Compensation** | `.Then.CompensatingAction` publishes the (previously orphaned) `Trigger…FailureHandlerCommand` → dedicated saga compensation chain | rollback runs once after exhaustion → `Failed` |
| DR-4 | **Timeout** | saga `TimeoutMessage` deadline race + idempotent guard | slow step routes to timeout; fast step no-ops |
| DR-5 | **Confidence** | saga `StepCompletedHandler` routes on `evt.Confidence` | 0.5 → `OnLowConfidence`; 0.9 → primary |
| DR-6 | **WithContext** | wire `ContextAssemblerEmitter` → ontology `IObjectSetProvider.ExecuteSimilarityAsync` (no `Strategos.Rag`) | step assembles state+retrieval+literal context |
| DR-7 | Expressibility | `Then<T>(configure)` on branch + failure-handler builders | IR-proven |
| DR-8 | Diagnostics | `AGWF017–021` (invalid retry/timeout/compensate/confidence) via the single-source TypeSpec catalog | 12 tests |
| DR-9 | **Behavioral harness** | new `Strategos.Generators.Behavioral.Tests` (Wolverine + Marten + Testcontainers Postgres) | the acceptance unlock |
| DR-10 | Composition | all four capabilities on one step; timeout/retry race; INV-7 | precedence + immutability proven on a real host |

## Why it matters

The new behavioral harness immediately caught **6 latent generator bugs** the old shape-only suite never could — including 3 that made the generated saga **fail to compile/run on Wolverine 6.x / Marten 9** (`[JasperFx.Identity]`, `long Version`, `WolverineFx.RuntimeCompilation`). Lowering is correctly **guarded**: a config-free step emits zero resilience artifacts (golden-baseline test).

## Verification

- Build: **33 projects, 0 errors, 0 warnings**
- `Strategos.Generators.Tests`: **1226 passed**
- `Strategos.Generators.Behavioral.Tests` (real Postgres): **17 passed**
- Spec + quality review: **APPROVED** (0 HIGH; the two MEDIUM findings — a latent CS0111 on compensation-step reuse, and a generic-type naming edge — were fixed with regression tests)

## Deferred follow-ons (documented in `docs/deferred-features.md`)

- [ ] `StepFailed`/`LowConfidenceRouted` **Marten stream events** (audit currently via queryable saga properties + logs; stream events apply only in EventSourced mode)
- [ ] Multi-step / rejoining `OnLowConfidence` handlers (single-step terminating handler lowered)
- [ ] Workflow-level `OnFailure` handler-chain interop with `Compensate<T>` (that chain was found independently dead)
- [ ] Validation-error-message literal escaping (pre-existing, not introduced here)

> **Note for reviewers:** `Refs` (not `Closes`) #135 — the epic has the tracked follow-ons above, so this PR delivers the core lowering without auto-closing the epic.

Refs #135

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/Strategos.Generators/WorkflowIncrementalGenerator.cs [hunk 1]
```diff
@@ -191,6 +204,41 @@ private static WorkflowGeneratorResult TransformToResult(
...
+                    mergedStepModels.Add(step with { Context = ctxModel });
+                }
+                else
+                {
+                    mergedStepModels.Add(step);
+                }
+            }
+
+            stepModels = mergedStepModels;
+        }
```
[the remark this answers is not in the document]
Comment (rsalus) [reply src/Strategos.Generators/WorkflowIncrementalGenerator.cs]: ✅ Fixed in `b11982d` — the context merge is deferred until after all step-model lowering (failure-handler + low-confidence handler steps appended), so `.WithContext` on those handler steps now attaches. Added `ContextOnHandlerStepsTests`.

File: src/Strategos.Generators/Helpers/StepExtractor.cs [hunk 2]
```diff
@@ -1123,16 +1226,262 @@ private static bool TryGetStepModel(
...
+                }
+                else if (compensation is null && SyntaxHelper.IsMethodCall(configCall, "Compensate"))
+                {
+                    compensation = ExtractCompensation(configCall, semanticModel);
+                }
+                else if (SyntaxHelper.IsMethodCall(configCall, "RequireConfidence")
+                    || SyntaxHelper.IsMethodCall(configCall, "OnLowConfidence"))
+                {
+                    confidence = MergeConfidence(confidence, configCall, semanticModel);
+                }
```
[the remark this answers is not in the document]
Comment (rsalus) [reply src/Strategos.Generators/Helpers/StepExtractor.cs]: ✅ Fixed in `bfb9539` — parent-step resilience extraction now walks only the current configure lambda's invocations (not nested-lambda descendants), so an inner `OnLowConfidence(alt => alt.Then<Y>(c => c.WithTimeout(t)))` no longer leaks onto the outer step. Covered by new `NestedLambdaLeakageTests`.

File: src/Strategos.Generators/Helpers/StepExtractor.cs [hunk 3]
```diff
@@ -1019,6 +1063,65 @@ private static void ParseForkPathStepModels(
...
+            foreach (var inv in pathInvocations)
+            {
+                // Scope the per-step validation lookup to this Then call's own configure lambda,
+                // identically to the fork-path parse, so branch-path steps lower their validation
+                // and resilience config the same way.
+                var (validationPredicate, validationErrorMessage) = ExtractConfiguredValidation(inv);
+
+                if (TryGetStepModel(inv, semanticModel, currentPrefix, validationPredicate, validationErrorMessage, out var stepModel))
+                {
+                    stepModels.Add(stepModel);
```
[the remark this answers is not in the document]
Comment (rsalus) [reply src/Strategos.Generators/Helpers/StepExtractor.cs]: ✅ Fixed in `bfb9539` — branch-path step discovery now uses a direct-`Then` walk (`CollectDirectThenInvocations`) that stops at nested-lambda boundaries, so nested handler `Then()` no longer surface as phantom branch steps.

File: src/Strategos.Generators/Emitters/Saga/SagaCompensationComponentEmitter.cs [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+                sb.AppendLine("        {");
+                sb.AppendLine($"            yield return new {workerCommandName}(WorkflowId, Guid.NewGuid(), State);");
+                sb.AppendLine("            yield break;");
+                sb.AppendLine("        }");
+                sb.AppendLine();
+            }
+        }
+
+        sb.AppendLine("    }");
+    }
```
[the remark this answers is not in the document]
Comment (rsalus) [reply src/Strategos.Generators/Emitters/Saga/SagaCompensationComponentEmitter.cs:211]: ✅ Fixed in `6de4f51` — multi-compensation routing now emits a terminal fallback (log → `Phase = Failed` → `MarkCompleted()` → `yield break`) after the `FailedStepName` branches, so an unmatched name can't strand the saga. Single-compensation always yields a command, so it's unaffected.

File: src/Strategos.Generators/Emitters/ExtensionsEmitter.cs [hunk 5]
```diff
@@ -186,6 +186,37 @@ private static void EmitAddMethod(StringBuilder sb, WorkflowModel model)
...
+                if (!emittedHeader)
+                {
+                    sb.AppendLine();
+                    sb.AppendLine("        // Register context assemblers");
+                    emittedHeader = true;
+                }
+
+                sb.AppendLine($"        services.AddTransient<{step.StepName}ContextAssembler>();");
+            }
+        }
```
[the remark this answers is not in the document]
Comment (rsalus) [reply src/Strategos.Generators/Emitters/ExtensionsEmitter.cs:275]: Investigated (`0a13ae1`): under current parse semantics branch/failure-handler steps are already folded into `model.Steps`, which the assembler-registration loop iterates — so once the merge-ordering fix (`b11982d`) lands, those steps **are** registered (verified with a probe + 6 passing `ContextOnHandlerStepsTests`, branch case included). Hardened the registration to mirror the branch/failure-handler sources idempotently anyway, to remove latent fragility if the model shape ever drifts.

File: src/Strategos.Generators/Emitters/Saga/StepCompletedHandlerEmitter.cs [hunk 6]
```diff
@@ -72,13 +72,25 @@ public void EmitHandler(
...
+        // own start command by the generator.
+        var confidence = context.StepModel?.Confidence;
         if (context.ApprovalAtStep is not null)
         {
             EmitApprovalWaitingHandler(sb, model, eventName, context.ApprovalAtStep);
         }
+        else if (confidence?.OnLowConfidenceHandlerStep is not null)
+        {
+            EmitConfidenceGatedHandler(sb, model, eventName, confidence, context);
+        }
```
[the remark this answers is not in the document]
Comment (rsalus) [reply src/Strategos.Generators/Emitters/Saga/StepCompletedHandlerEmitter.cs:93]: ✅ Fixed in `add2608` — the confidence-gated handler now emits the `Phase == Failed` route to the failure-handler start command **before** the confidence comparison (gated on `HasFailureHandlers`, mirroring `EmitPhaseAwareNonFinalStepHandler`). The no-failure-handler path is byte-unchanged.

File: src/Strategos.Generators/Helpers/FailureHandlerExtractor.cs [hunk 7]
```diff
@@ -182,6 +182,17 @@ private static bool TryGetStepNameAndModel(
...
+        // Prefer the shared configured-step builder so any per-step resilience
+        // (WithRetry/WithTimeout/Compensate/confidence) and ValidateState guard declared via the
+        // Then<TStep>(step => step...) configure-lambda overload (DR-7) threads into the StepModel,
+        // bringing failure-handler steps to parity with the top-level/loop/fork parse paths.
+        if (StepExtractor.TryBuildConfiguredStepModel(invocation, semanticModel, out var configuredStepModel))
+        {
+            stepName = configuredStepModel.StepName;
+            stepModel = configuredStepModel;
+            return true;
+        }
```
[the remark this answers is not in the document]
Comment (rsalus) [reply src/Strategos.Generators/Helpers/FailureHandlerExtractor.cs:193]: ✅ Fixed in `bfb9539` — failure-handler step extraction now uses `InvocationChainWalker.CollectInvocationsInLambda` (stops at nested-lambda boundaries), so nested `Then()` inside configure lambdas (e.g. `OnLowConfidence`) are no longer captured as failure-handler steps.

File: src/Strategos.Generators/Helpers/ResilienceParser.cs [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+        if (expression is LiteralExpressionSyntax literal
+            && literal.IsKind(SyntaxKind.NumericLiteralExpression)
+            && literal.Token.Value is int intValue)
+        {
+            value = intValue;
+            return true;
+        }
+
+        return false;
+    }
```
[the remark this answers is not in the document]
Comment (rsalus) [reply src/Strategos.Generators/Helpers/ResilienceParser.cs:129]: ✅ Fixed in `3d47cf4` — the retry-count parser now handles `PrefixUnaryExpressionSyntax`, so `WithRetry(-1)` reaches the IR and `RetryMaxAttemptsBelowOne` (AGWF020) fires. Added a diagnostics test for the negative literal.

File: src/Strategos.Generators/Helpers/ResilienceParser.cs [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+                return true;
+            case "FromHours":
+                value = TimeSpan.FromHours(amount);
+                return true;
+            case "FromDays":
+                value = TimeSpan.FromDays(amount);
+                return true;
+            default:
+                return false;
+        }
```
[the remark this answers is not in the document]
Comment (rsalus) [reply src/Strategos.Generators/Helpers/ResilienceParser.cs]: ✅ Fixed in `b62b7ee` — the `TimeSpan.From*` construction is wrapped to return `false` on `OverflowException`/`ArgumentException` instead of crashing generation on an extreme literal.

File: src/Strategos.Generators.Behavioral.Tests/Infrastructure/ContextProbe.cs [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    /// </summary>
+    public int SimilarityCallCount { get; private set; }
+
+    /// <summary>
+    /// Gets the assembled context the generated worker handler delivered to the
+    /// context-aware step, or <see langword="null"/> if the step has not received
+    /// context since the last <see cref="Reset"/>.
+    /// </summary>
+    public AssembledContext? LastAssembledContext { get; private set; }
+
```
[the remark this answers is not in the document]
Comment (rsalus) [reply src/Strategos.Generators.Behavioral.Tests/Infrastructure/ContextProbe.cs]: ✅ Fixed in `921465e` — the three observation getters now acquire the same `gate` lock as the writers (backing fields), so a reader can't observe torn/stale state on this shared DI singleton.

File: src/Strategos.Generators.Behavioral.Tests/Infrastructure/PostgresFixture.cs [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    private static void ConfigurePodmanProvider()
+    {
+        SetIfAbsent("DOCKER_HOST", DefaultPodmanSocket);
+
+        // Ryuk (the Testcontainers resource reaper) commonly fails to start
+        // under rootless podman; disable it so the run is not blocked. The
+        // container's own WithCleanUp(true) still removes the container on
+        // dispose.
+        SetIfAbsent("TESTCONTAINERS_RYUK_DISABLED", "true");
+    }
```
[the remark this answers is not in the document]
Comment (rsalus) [reply src/Strategos.Generators.Behavioral.Tests/Infrastructure/PostgresFixture.cs:146]: ✅ Fixed in `d4d940f` — podman self-config is now guarded three ways: skip if `DOCKER_HOST` is already set, only apply when the rootless podman socket actually exists on disk, and only on Linux. A Docker-only runner using default discovery is left untouched.

File: src/Strategos.Generators.Behavioral.Tests/Workflows/CompositionEdgeWorkflow.cs [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+    public void RecordInput(CompositionState input)
+    {
+        ArgumentNullException.ThrowIfNull(input, nameof(input));
+        this.seenInputs.Enqueue(input);
+    }
+
+    /// <summary>
+    /// Clears the probe at the start of each test to isolate it from prior runs.
+    /// </summary>
+    public void Reset() => this.seenInputs.Clear();
```
[the remark this answers is not in the document]
Comment (rsalus) [reply src/Strategos.Generators.Behavioral.Tests/Workflows/CompositionEdgeWorkflow.cs]: ✅ Fixed in `1b765c1` — attempts are now tracked per-run in `CompositionImmutableProbe` keyed by `WorkflowId` (stable across retries, which is exactly what T021 asserts), not the per-session shared log. Behavioral suite ran green twice back-to-back to confirm order-independence.

**Comment by rsalus** [comment]
## CodeRabbit review — all 13 findings addressed ✅

Thanks for the thorough pass — these were genuinely valuable edge cases. Pushed in `707cb50`. **12 were real bugs (fixed with regression tests); 1 was a false positive under the current parse semantics but hardened anyway.** Each inline thread has a specific reply + commit.

| Area | Finding | Resolution |
|------|---------|-----------|
| Parse | Nested-lambda leakage (resilience config / steps from inner `OnLowConfidence`/configure lambdas) | `bfb9539` — `CollectInvocationsInLambda` + new `CollectDirectThenInvocations` across the 5 fork/branch walks; relocates legit nested config to the right extractor (`NestedLambdaLeakageTests`) |
| Parse | `WithRetry(-n)` silently dropped | `3d47cf4` — parse unary-minus → `AGWF020` now fires |
| Parse | `TimeSpan.From*` overflow crashes generation | `b62b7ee` — guarded → returns `false` |
| Emit | Confidence gate bypasses `Phase==Failed` routing | `add2608` — emit the failed-phase route (gated on `HasFailureHandlers`) |
| Emit | Compensation dead-end on unmatched `FailedStepName` | `6de4f51` — terminal fallback (no more saga hang) |
| Emit | Context merge ordering loses `.WithContext` on handler steps | `b11982d` — defer merge until after all step-model lowering |
| Emit | Assemblers for branch/failure steps | `0a13ae1` — already covered via fold-into-`Steps` (verified); hardened idempotently |
| Model | `WorkflowModel.Create` drops `ConfidenceHandlerStepNames` | `6179d56` — threaded through `Create` (hot path uses the ctor, so low impact; factory now consistent) |
| Test | `ContextProbe` unsynchronized reads | `921465e` — lock-guarded getters |
| Test | `PostgresFixture` podman config breaks Docker-only CI | `d4d940f` — socket-exists + Linux guards |
| Test | Composition retry test driven by shared log → flaky | `1b765c1` — per-run attempt tracking keyed by `WorkflowId` |

**CI:** the `coverage-gate` failure was a measurement artifact — the new behavioral-test project pulls in WolverineFx/Marten, whose runtime-codegen core **JasperFx (~48k instrumented lines @ 29%)** was counted as "production", dragging the weighted aggregate from ~82% to 55.15%. Excluded the third-party runtime assemblies from the coverage filter (`a6400dc`); Strategos-only production coverage is ~82.5% (verified from the failing run's cobertura report).

Local verification on the pushed tip: build **0/0**, generator suite **1243/1243**, behavioral suite **17/17** on a real Postgres container.
